src/openms/source/ANALYSIS/MAPMATCHING/QTClusterFinder.cpp

```diff
@@ -294,50 +301,72 @@ namespace OpenMS
     if (do_progress) logger.endProgress();
   }
 
-  void QTClusterFinder::makeConsensusFeature_(list<QTCluster>& clustering,
+  bool QTClusterFinder::makeConsensusFeature_(Heap& cluster_heads,
                                               ConsensusFeature& feature,
                                               ElementMapping& element_mapping,
-                                              Grid& grid)
+                                              Grid& grid,
+                                              const vector<Handle>& handles)
   {
-    // find the best cluster (a valid cluster with the highest score)
-    // -> this is equivalent to std::max_element but we can skip invalid clusters
-    list<QTCluster>::iterator best = clustering.begin();
-    while (best != clustering.end() && best->isInvalid()) // find start element
+    // pop until the top is valid
+    while (cluster_heads.top().isInvalid())
     {
-      ++best;
+      removeTopFromHeap_(cluster_heads, cluster_heads.top(), element_mapping);
+      
+      // if the last remaining cluster was invalid, no consensus feature is created
+      if (cluster_heads.empty()) return false;
     }
-    for (list<QTCluster>::iterator it = best;
-         it != clustering.end(); ++it)
+
+    const QTCluster& best = cluster_heads.top();
+
+    ClusterElementsMap const elements = best.getElements();
+
+    #ifdef DEBUG_QTCLUSTERFINDER
+    std::cout << "Elements: " << elements.size() << " with best "
+         << best->getQuality() << " invalid " << best->isInvalid() << std::endl;
+    #endif
+
+    createConsensusFeature_(feature, best.getCurrentQuality(), elements);
+
+    #ifdef DEBUG_QTCLUSTERFINDER
+    std::cout << " create new consensus feature " << feature.getRT() << " " << feature.getMZ() << " from " << best->getCenterPoint()->getFeature().getUniqueId() << std::endl;
+    for (OpenMSBoost::unordered_map<Size, OpenMS::GridFeature*>::const_iterator
+         it = elements.begin(); it != elements.end(); ++it)
     {
-      if (!it->isInvalid())
-      {
-        if (it->getQuality() > best->getQuality())
-        {
-          best = it;
-        }
-      }
+      std::cout << "   = element id : " << it->second->getFeature().getUniqueId() << std::endl;
     }
+    #endif
+
+    updateClustering_(element_mapping, grid, elements, cluster_heads, handles, best.getId());
 
-    // no more clusters to process -> clear clustering and return
-    if (best == clustering.end())
+    return true;
+  }
+
+  void QTClusterFinder::removeTopFromHeap_(Heap& cluster_heads,
+                                           const QTCluster& cluster,
+                                           ElementMapping& element_mapping)
+  {
+    ClusterElementsMap const elements = cluster.getElements();
```

Comment:
how about:
```
// some useful comment here
for (const auto& elements : cluster.getElements())
{
  element_mapping[elements.second.second].erase(cluster.getId());
}

cluster_heads.pop();
```

Reply:
`elements.second.second` indicates a rather complicated structure... maybe replace `pair` with a struct, so you can use proper names?

Reply:
Done, I think it improved readabilty a lot.